blockchain/indexers/slpindex.go

```diff
@@ -0,0 +1,877 @@
+		entry := &dbSlpIndexEntry{
+			tx:             tx,
+			slpMsg:         slpMsg,
+			tokenIDHash:    tokenIDHash,
+			slpMsgPkScript: tx.TxOut[0].PkScript,
+			slpVersion:     uint16(slpMsg.TokenType),
+		}
+		return dbPutSlpIndexEntry(idx, dbTx, entry)
+	}
+
+	burnedInputs := make([]*BurnedInput, 0)
+	for _, tx := range sortedTxns {
+		isValid, txnInputsBurned := CheckSlpTx(tx, getSlpIndexEntry, putTxIndexEntry)
+
+		// look for burned inputs within non-SLP txns
+		if !isValid {
+			for _, txi := range tx.TxIn {
+				slpEntry, _ := idx.GetSlpIndexEntry(dbTx, &txi.PreviousOutPoint.Hash)
+				if slpEntry != nil {
+					slpMsg, _ := v1parser.ParseSLP(slpEntry.SlpOpReturn)
+					burnedInputs = append(burnedInputs, &BurnedInput{
+						Tx:      tx,
+						TxInput: txi,
+						SlpMsg:  slpMsg,
+						Entry:   slpEntry,
+					})
+				}
+			}
+		}
+
+		if txnInputsBurned != nil {
+			burnedInputs = append(burnedInputs, txnInputsBurned...)
+		}
+	}
+
+	// Loop through burned inputs and check for different situations
+	// where token metadata will need to be updated.
+	//
+	// NOTE: items in burnedInputs are not topologically ordered
+	//
+	for _, burn := range burnedInputs {
+		// Currently we only need to check for a burned mint baton
+		isMintBatonBurned := idx.checkBurnedInputForMintBaton(dbTx, burn)
+		if isMintBatonBurned {
+			continue
+		}
+	}
+	return nil
+}
+
+func (idx *SlpIndex) checkBurnedInputForMintBaton(dbTx database.Tx, burn *BurnedInput) bool {
+
+	// we can skip nft children since they don't have mint batons
+	if burn.SlpMsg.TokenType == 0x41 {
+		return false
+	}
+
+	// check if input is the mint baton from either Genesis or Mint parent data
+	if msg, ok := burn.SlpMsg.Data.(v1parser.SlpGenesis); ok {
+		if msg.MintBatonVout != int(burn.TxInput.PreviousOutPoint.Index) {
+			return false
+		}
+	} else if msg, ok := burn.SlpMsg.Data.(v1parser.SlpMint); ok {
+		if msg.MintBatonVout != int(burn.TxInput.PreviousOutPoint.Index) {
+			return false
+		}
+	} else {
+		return false
+	}
+
+	// double-check this burned mint baton was a valid slp token
+	if burn.Entry == nil {
+		return false
+	}
+
+	err := dbPutTokenIDIndexEntry(dbTx, burn.Entry.TokenID,
+		&TokenMetadata{
+			TokenID:       &burn.Entry.TokenIDHash,
+			SlpVersion:    burn.Entry.SlpVersionType,
+			MintBatonHash: nil,
+			MintBatonVout: 0,
+			NftGroupID:    nil,
+		},
+	)
+	if err != nil {
+		panic("could not update token metadata")
+	}
+
+	return true
+}
+
+// GetSlpIndexEntryHandler ...
+type GetSlpIndexEntryHandler func(*chainhash.Hash) (*SlpIndexEntry, error)
+
+// AddTxIndexEntryHandler ...
+type AddTxIndexEntryHandler func(*wire.MsgTx, *v1parser.ParseResult, *chainhash.Hash) error
+
+// CheckSlpTx checks a transaction for validity and adds valid txns to the db
+func CheckSlpTx(tx *wire.MsgTx, getSlpIndexEntry GetSlpIndexEntryHandler, putTxIndexEntry AddTxIndexEntryHandler) (bool, []*BurnedInput) {
+
+	txSlpMsg, _ := v1parser.ParseSLP(tx.TxOut[0].PkScript)
```

Comment:
NIT: do we need a length check on outputs?

Reply:
done. Also handled other cases of the same thing.